# Proyecto Final Analítica de Datos

#### Juan Diego Carvajal Castaño - A00135764
#### Juan Manuel Imbachí Güengue - A00130580

Importamos las librerías que vamos a usar

In [ ]:
import numpy as np #operaciones matriciales y con vectores
import pandas as pd #tratamiento de datos
import matplotlib.pyplot as plt #gráficos

In [ ]:
data = pd.read_csv('2017-18_NBA_salary.csv')

## Preparación de los datos

Info de los datos -- si se quiere

In [63]:
data.shape

(483, 28)

In [64]:
data.head()

Player    Salary NBA_Country  NBA_DraftNumber  Age   Tm   G    MP  \
0        Zhou Qi    815615       China               43   22  HOU  16    87   
1  Zaza Pachulia   3477600     Georgia               42   33  GSW  66   937   
2  Zach Randolph  12307692         USA               19   36  SAC  59  1508   
3    Zach LaVine   3202217         USA               13   22  CHI  24   656   
4   Zach Collins   3057240         USA               10   20  POR  62   979   

    PER    TS%   3PAr    FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  \
0   0.6  0.303  0.593  0.370   6.5  16.8  11.7   1.5   1.1   6.8  18.2  19.5   
1  16.8  0.608  0.004  0.337  11.0  25.0  18.5  15.4   1.9   1.3  19.3  17.2   
2  17.3  0.529  0.193  0.140   7.0  23.8  15.0  14.9   1.4   0.6  12.5  27.6   
3  14.6  0.499  0.346  0.301   1.4  14.4   7.7  18.6   1.8   0.5   9.7  29.5   
4   8.2  0.487  0.387  0.146   4.9  18.3  11.7   7.3   0.8   2.5  15.6  15.5   

   OWS  DWS   WS  WS/48  OBPM  DBPM   BPM  VORP  
0 -0.4  0.1 -0.2 -0.121 -10.6   0.5 -10.1  -0.2  
1  1.7  1.4  3.1  0.160  -0.6   1.3   0.8   0.7  
2  0.3  1.1  1.4  0.046  -0.6  -1.3  -1.9   0.0  
3 -0.1  0.5  0.4  0.027  -0.7  -2.0  -2.6  -0.1  
4 -0.4  1.2  0.8  0.038  -3.7   0.9  -2.9  -0.2

Hay 485 registros con 28 variables.
Ahora analizaremos si hay problemas de datos con respecto a valores faltantes y outliers.

In [47]:
data.describe(include="all").transpose()

count unique         top freq         mean          std  \
Player            483    481  Kay Felder    3          NaN          NaN   
Salary            483    NaN         NaN  NaN  6.66062e+06  7.39821e+06   
NBA_Country       483     44         USA  372          NaN          NaN   
NBA_DraftNumber   483    NaN         NaN  NaN      29.3954       21.116   
Age               483    NaN         NaN  NaN      26.2733      4.27429   
Tm                483     31         TOT   55          NaN          NaN   
G                 483    NaN         NaN  NaN      50.3706      24.7236   
MP                483    NaN         NaN  NaN      1158.92      809.628   
PER               483    NaN         NaN  NaN      13.3157      8.74567   
TS%               483    NaN         NaN  NaN     0.535387     0.112352   
3PAr              483    NaN         NaN  NaN     0.337383     0.226894   
FTr               483    NaN         NaN  NaN     0.263404     0.294578   
ORB%              483    NaN         NaN  NaN        4.894      4.58152   
DRB%              483    NaN         NaN  NaN      15.0126      6.79352   
TRB%              483    NaN         NaN  NaN      9.94928      4.92536   
AST%              483    NaN         NaN  NaN      13.0014      9.09296   
STL%              483    NaN         NaN  NaN      1.53582     0.986686   
BLK%              483    NaN         NaN  NaN      1.72029      1.68365   
TOV%              483    NaN         NaN  NaN      13.1404      6.11529   
USG%              483    NaN         NaN  NaN      18.9756      5.82653   
OWS               483    NaN         NaN  NaN      1.28075      1.88354   
DWS               483    NaN         NaN  NaN      1.18137      1.03394   
WS                483    NaN         NaN  NaN      2.46542      2.67452   
WS/48             483    NaN         NaN  NaN    0.0803251     0.163229   
OBPM              483    NaN         NaN  NaN     -1.25259      5.02875   
DBPM              483    NaN         NaN  NaN    -0.489441      2.39416   
BPM               483    NaN         NaN  NaN     -1.74182      5.66591   
VORP              483    NaN         NaN  NaN     0.601242      1.24764   

                   min          25%          50%     75%          max  
Player             NaN          NaN          NaN     NaN          NaN  
Salary           46080  1.47138e+06  3.21828e+06   1e+07  3.46826e+07  
NBA_Country        NaN          NaN          NaN     NaN          NaN  
NBA_DraftNumber      1         10.5           25      47           62  
Age                 19           23           26      29           41  
Tm                 NaN          NaN          NaN     NaN          NaN  
G                    1         29.5           59      71           79  
MP                   1          385         1151  1824.5         2898  
PER              -41.1          9.8         13.2   16.55        134.1  
TS%                  0       0.5055        0.545  0.5825          1.5  
3PAr                 0        0.167        0.346   0.481            1  
FTr                  0        0.155        0.231  0.3195        5.333  
ORB%                 0          1.8          3.3    7.05         35.9  
DRB%                 0         10.3           14    18.8         37.6  
TRB%                 0          6.2          8.7    13.3         26.5  
AST%                 0          6.9           10    17.6         49.4  
STL%                 0            1          1.5     1.9         12.5  
BLK%                 0          0.6          1.2     2.2         13.4  
TOV%                 0          9.9         12.5   15.75         66.7  
USG%               5.7           15         17.9   22.25         45.1  
OWS               -2.3            0          0.8       2         11.4  
DWS                  0          0.3            1     1.8          5.6  
WS                -1.2          0.3          1.9     3.6           15  
WS/48           -1.063       0.0405        0.083   0.123        2.713  
OBPM             -36.5         -2.6     

Aparentemente no hay outliers, sin embargo, se puede observar que hay valores faltantes, puesto que el valor de la columna COUNT no es 485 para todas las variables.

### Variables que tienen valores faltantes


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485 entries, 0 to 484
Data columns (total 28 columns):
Player             485 non-null object
Salary             485 non-null int64
NBA_Country        485 non-null object
NBA_DraftNumber    485 non-null int64
Age                485 non-null int64
Tm                 485 non-null object
G                  485 non-null int64
MP                 485 non-null int64
PER                485 non-null float64
TS%                483 non-null float64
3PAr               483 non-null float64
FTr                483 non-null float64
ORB%               485 non-null float64
DRB%               485 non-null float64
TRB%               485 non-null float64
AST%               485 non-null float64
STL%               485 non-null float64
BLK%               485 non-null float64
TOV%               483 non-null float64
USG%               485 non-null float64
OWS                485 non-null float64
DWS                485 non-null float64
WS                 485 non-n

In [15]:
print(pd.isnull(data).any(0)) #el 0 implica cualquier columna, si quisieramos las filas utilizamos 1
print("Las columnas que tienen valores faltantes son: ",pd.isnull(data).any(0).to_numpy().nonzero()[0])
print("y corresponden a: ", data.columns[pd.isnull(data).any(0).to_numpy().nonzero()[0]])

Player             False
Salary             False
NBA_Country        False
NBA_DraftNumber    False
Age                False
Tm                 False
G                  False
MP                 False
PER                False
TS%                 True
3PAr                True
FTr                 True
ORB%               False
DRB%               False
TRB%               False
AST%               False
STL%               False
BLK%               False
TOV%                True
USG%               False
OWS                False
DWS                False
WS                 False
WS/48              False
OBPM               False
DBPM               False
BPM                False
VORP               False
dtype: bool
Las columnas que tienen valores faltantes son:  [ 9 10 11 18]
y corresponden a:  Index(['TS%', '3PAr', 'FTr', 'TOV%'], dtype='object')


#### Las columnas TS%, 3PAr, FTr, TOV% tienen valores faltantes. 

### Registros repetidos

In [33]:
data['Player'].value_counts()

Kay Felder                  3
Denzel Valentine            1
Aaron Gordon                1
Kentavious Caldwell-Pope    1
Elfrid Payton               1
Milos Teodosic              1
Nicolas Brussino            1
Tony Bradley                1
Larry Drew                  1
Meyers Leonard              1
Evan Turner                 1
Georgios Papagiannis        1
Ike Anigbogu                1
Alfonzo McKinnie            1
Kyle Lowry                  1
Austin Rivers               1
Nikola Jokic                1
Al-Farouq Aminu             1
Devin Harris                1
Darrell Arthur              1
Tyler Zeller                1
Jameel Warney               1
Sterling Brown              1
Nigel Hayes                 1
Michael Beasley             1
Devin Booker                1
D.J. Wilson                 1
James Ennis                 1
Sam Dekker                  1
Goran Dragic                1
                           ..
Patty Mills                 1
Tyler Ennis                 1
Pat Connau

In [41]:
data[data['Player'].duplicated(keep=False)]

Player   Salary NBA_Country  NBA_DraftNumber  Age   Tm   G   MP  \
224  Kay Felder  1312611         USA               54   22  DET   1    3   
225  Kay Felder  1312611         USA               54   22  CHI  14  134   
226  Kay Felder  1312611         USA               54   22  TOT  15  137   

      PER    TS%  ...  TOV%  USG%  OWS  DWS   WS  WS/48  OBPM  DBPM   BPM  \
224 -31.6  0.000  ...  33.3  44.2 -0.1  0.0 -0.1 -1.005 -29.5 -11.9 -41.4   
225   3.6  0.386  ...  17.4  28.0 -0.5  0.1 -0.5 -0.166  -8.2  -3.3 -11.5   
226   2.9  0.375  ...  17.9  28.4 -0.6  0.1 -0.5 -0.185  -8.7  -3.5 -12.1   

     VORP  
224   0.0  
225  -0.3  
226  -0.3  

[3 rows x 28 columns]

#### Se encontró que el jugador Kay Felder tiene 3 registros asociados a 3 equipos diferentes durante la temporada

# Reporte Primera Entrega

## Calidad de los Datos

Al observar los datos se pudo evidenciar que:
### Faltantes
   Existen datos faltantes para 4 variables: TS%, 3PAr, FTr, TOV%, donde cada una tiene 2 valores nulos o NaN.
### Outliers
   Columna OWS tiene valores negativos pero representa el número de victorias a las que contribuyó el jugador dado su desempeño ofensivo.
    WS,WS/48,OBPM,DBPM,BPM,VORP son negativos, ¿tiene sentido?

### Repetidos
   Usando el atributo 'Player' como identificador del registro, se encontró que hay 3 registros repetidos con el valor'Kay Felder', aunque sabemos que es inusual que un jugador sea transferido más de una vez dentro de una misma temporada, dado que es posible realizar dicha transferencia, tomaremos estos 3 registros en el análisis.


### Problemas de escala  
* En el caso de los porcentajes, las variables varían en el orden de magnitud, unas presentan un orden de 10^0 y otras de 10^-1.
* Las variables presentan diferencias significativas en sus promedios, valores mínimos y máximos.
    
   Estos problemas de escala pueden ser consecuencia de valores atípicos o outliers que están presentes en los datos.

### Problemas de tipo
* Las características 'NBA_Country', 'Tm' y 'Player' son las únicas de tipo string, por lo que deberían ser convertidas a variables numéricas para ser tomadas en cuenta en el estudio. Sin embargo, consideramos que 'Player' no es relevante para hacer inferencias sobre el posible salario de un jugador.
    
### Distribución de los datos
    Los valores dentro de cada uno de los percentiles varía significativamente,
    
#### __*El registro 142 tiene valores muy elevados*__
#### __*Los registros 147 y 224 tienes valores muy bajos*__

## Eliminación de registros con valores faltantes

In [18]:
cleanData = data.dropna()
print("Datos originales: ", data.shape)
print("Datos limpios y completos: ", cleanData.shape)
print("Se eliminaron {} registos".format(data.shape[0] - cleanData.shape[0]))

Datos originales:  (485, 28)
Datos limpios y completos:  (483, 28)
Se eliminaron 2 registos


In [19]:
data = cleanData

## Eliminación de registros repetidos

In [46]:
data.drop_duplicates(subset=['Player'])['Player'].value_counts() #Parámetro keep para definir si dejar alguno de los duplicados first/last

Denzel Valentine            1
Cristiano Felicio           1
Kentavious Caldwell-Pope    1
Elfrid Payton               1
Milos Teodosic              1
Nicolas Brussino            1
Tony Bradley                1
Larry Drew                  1
Meyers Leonard              1
Evan Turner                 1
Georgios Papagiannis        1
Ike Anigbogu                1
Alfonzo McKinnie            1
Kyle Lowry                  1
Austin Rivers               1
Aaron Gordon                1
Nikola Jokic                1
Devin Harris                1
Darrell Arthur              1
Tyler Zeller                1
Jameel Warney               1
Sterling Brown              1
Nigel Hayes                 1
Michael Beasley             1
Devin Booker                1
D.J. Wilson                 1
James Ennis                 1
Sam Dekker                  1
Goran Dragic                1
Willie Cauley-Stein         1
                           ..
Patty Mills                 1
Tyler Ennis                 1
Pat Connau

# Maousekeherramientas para usar más adelante


In [66]:
# Filtrado
data['WS/48'].sort_values(ascending=True)

147   -1.063
224   -1.005
188   -0.443
6     -0.251
84    -0.237
305   -0.211
161   -0.196
375   -0.192
241   -0.189
226   -0.185
135   -0.180
7     -0.169
225   -0.166
143   -0.141
289   -0.141
134   -0.134
0     -0.121
245   -0.118
22    -0.103
449   -0.080
196   -0.063
94    -0.060
391   -0.060
23    -0.057
276   -0.053
393   -0.047
110   -0.046
280   -0.043
167   -0.036
302   -0.036
341   -0.036
313   -0.035
112   -0.033
321   -0.033
420   -0.033
202   -0.030
274   -0.030
180   -0.027
333   -0.020
83    -0.020
185   -0.019
410   -0.019
184   -0.018
381   -0.017
242   -0.016
392   -0.015
239   -0.015
437   -0.012
399   -0.012
28    -0.011
377   -0.010
432   -0.010
285   -0.009
361   -0.008
282   -0.008
363   -0.006
317   -0.006
456   -0.005
478   -0.004
460   -0.001
352    0.000
481    0.004
365    0.005
125    0.006
369    0.006
458    0.006
314    0.008
334    0.008
32     0.008
72     0.010
454    0.013
47     0.016
152    0.016
407    0.017
370    0.018
168    0.018
482    0.019

In [62]:
# Obtener registro por index
pd.options.display.max_columns = None
pd.options.display.max_rows = None
data.iloc[140:141]
data.iloc[145:146]

Player   Salary NBA_Country  NBA_DraftNumber  Age   Tm  G  \
147  Mindaugas Kuzminskas  3025035   Lithuania               62   28  NYK  1   

     MP   PER  TS%  3PAr  FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  \
147   2 -41.1  0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  43.6   

     OWS  DWS   WS  WS/48  OBPM  DBPM   BPM  VORP  
147  0.0  0.0  0.0 -1.063 -36.5 -12.7 -49.2   0.0